tests/test_device_validator.py

```diff
@@ -0,0 +1,136 @@
+            validate_load_device(fit_to_limit=True,
+                                 energy_rate_increase_per_update=2)
+
+    def test_pv_device_setting(self):
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(panel_count=-1)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(final_selling_rate=-1)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(initial_selling_rate=-20)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(initial_selling_rate=10,
+                               final_selling_rate=11)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(fit_to_limit=True,
+                               energy_rate_decrease_per_update=2)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(energy_rate_decrease_per_update=-1)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_pv_device(max_panel_power_W=-5)
+
+    def test_storage_device_setting(self):
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(initial_soc=5)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(min_allowed_soc=2)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(initial_soc=15, min_allowed_soc=20)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(battery_capacity_kWh=-1)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(max_abs_battery_power_kW=-1)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(initial_selling_rate=-1)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(final_selling_rate=-2)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(initial_selling_rate=10,
+                                    final_selling_rate=11)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(initial_buying_rate=-2)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(final_buying_rate=-3)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(initial_buying_rate=10,
+                                    final_buying_rate=9)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(final_buying_rate=15,
+                                    final_selling_rate=14)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(energy_rate_increase_per_update=-5)
+
+        with self.assertRaises(D3ADeviceException):
+            validate_storage_device(energy_rate_decrease_per_update=-3)
+
+        with self.assertRaises(D3ADeviceException):
```

Comment:
@faizan2590 Can you please add a test for the case, where it shouldn't raise an Exception?
Thanks!

Reply:
Okay, doing it!